components/utilities/ai.js

```diff
@@ -0,0 +1,184 @@
+        for (let j = i + 1; j < dupesObject[value].length; j++) {
+          combinations.push([dupesObject[value][i], dupesObject[value][j]]);
+        }
+      }
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a triple */
+/** 
+ * NOTE: This solution for if the number of duplicates is four seems severely need of refactoring.
+*/
+const triple = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const combinations = [];
+
+  for (const value in dupesObject) {
+    if (dupesObject[value].length === 3) {
+      combinations.push(dupesObject[value]);
+    } else if (dupesObject[value].length === 4) {
+      for (let i = 0; i < dupesObject[value].length; i++) {
+        for (let j = i + 1; j < dupesObject[value].length; j++) {
+          if (dupesObject[value][j + 1]) {
+            combinations.push([dupesObject[value][i], dupesObject[value][j], dupesObject[value][j + 1]]);
+          }
+
+          if (dupesObject[value][j + 2]) {
+            combinations.push([dupesObject[value][i], dupesObject[value][j], dupesObject[value][j + 2]]);
+          }
+        }
+      }
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a quartet */
+const quartet = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const combinations = [];
+
+  for (const value in dupesObject) {
+    if (dupesObject[value].length === 4) {
+      combinations.push(dupesObject[value]);
+    }
+  }
+  return combinations;
+}
+
+/* AI logic to play double sequence */
+
+
+const merge = (left, right) => {
+  let sortedArr = [];
+
+  while (left.length && right.length) {
+    if (left[0].value < right[0].value) {
+      sortedArr.push(left.shift());
+    } else {
+      sortedArr.push(right.shift());
+    }
+  }
+
+  return [...sortedArr, ...left, ...right];
+}
+
+const mergeSort = (arr) => {
+  if (arr.length <= 1) {
+    return arr;
+  }
+
+  const mid = Math.floor(arr.length / 2);
+
+  const left = mergeSort(arr.slice(0, mid));
+  const right = mergeSort(arr.slice(mid));
+
+  return merge(left, right);
+}
+
+/*
+'hearts': '♥',
+'diamonds': '♦',
+'spades': '♠',
+'clubs': '♣',
+*/
+
+const createDuplicatesObject = (array) => {
+  const values = {};
+
+  for (let i = 0; i < array.length; i++) {
+    const number = array[i].number;
+    if (!values[number]) {
+      values[number] = [array[i]];
+    } else {
+      values[number].push(array[i]);
+    }
+  }
+  return values;
+}
+
+// Duplicates Object example
+/*
+ {
+'6': [ 'hearts' ],
+'7': [ 'clubs' ],
+'8': [ 'diamonds' ],
+'9': [ 'hearts', 'clubs' ],
+'10': [ 'hearts' ],
+'11': [ 'hearts' ],
+'12': [ 'hearts' ],
+'13': [ 'hearts' ],
+'14': [ 'hearts' ],
+'15': [ 'clubs', 'diamonds' ]
+}
+*/
+/* AI logic to play a sequence */
+const sequence = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const sortedHand = mergeSort(hand);
+  const totalSequences = [];
+
+  for (let i = 0; i < sortedHand.length; i++) {
```

Comment:
i'm going to need to take another look at this. and probably also at the functions above as well. you have a lot of for loops nested within for loops, and i think some of these can probably be re-written as reduce functions to make this a bit simpler to read and possibly even run faster.

Reply:
Leaving this comment so you can take a look again. I'll take a look when I have free time.